src/ray/id.h

```diff
@@ -102,13 +151,78 @@ const TaskID ComputeTaskId(const ObjectID &object_id);
 const TaskID GenerateTaskId(const DriverID &driver_id, const TaskID &parent_task_id,
                             int parent_task_counter);
 
-/// Compute the index of this object in the task that created it.
-///
-/// \param object_id The object ID.
-/// \return The index of object creation according to the task that created
-/// this object. This is positive if the task returned the object and negative
-/// if created by a put.
-int64_t ComputeObjectIndex(const ObjectID &object_id);
+template <typename T>
+T BaseID<T>::from_random() {
+  std::string data(T::size(), 0);
+  // NOTE(pcm): The right way to do this is to have one std::mt19937 per
+  // thread (using the thread_local keyword), but that's not supported on
+  // older versions of macOS (see https://stackoverflow.com/a/29929949)
+  static std::mutex random_engine_mutex;
+  std::lock_guard<std::mutex> lock(random_engine_mutex);
+  static std::mt19937 generator = RandomlySeededMersenneTwister();
+  std::uniform_int_distribution<uint32_t> dist(0, std::numeric_limits<uint8_t>::max());
+  for (int i = 0; i < T::size(); i++) {
+    data[i] = static_cast<uint8_t>(dist(generator));
+  }
+  return T::from_binary(data);
+}
+
+template <typename T>
+T BaseID<T>::from_binary(const std::string &binary) {
+  T t = T::nil();
+  std::memcpy(reinterpret_cast<uint8_t *>(&t), binary.data(), T::size());
+  return t;
+}
+
+template <typename T>
+const T &BaseID<T>::nil() {
+  static const T nil_id;
+  return nil_id;
+}
+
+template <typename T>
+bool BaseID<T>::is_nil() const {
+  const uint8_t *d = data();
+  for (int i = 0; i < T::size(); ++i) {
+    if (d[i] != 0xff) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename T>
+bool BaseID<T>::operator==(const BaseID &rhs) const {
+  return std::memcmp(data(), rhs.data(), T::size()) == 0;
```

Comment:
compare hash code first

Reply:
`MurmurHash64A` is expensive. It is faster to compare the bytes directly and delay the hash calculation.

Reply:
is this still the case given that the hash code is computed only once?
or maybe you can instead check if the hash code exists, if it does, match the hash code first?

Reply:
Currently it is calculated only once. The `MurmurHash64A` is expensive that is why we change it to lazy evaluation.  I think directly comparison is much faster than calculating the hash value. Moreover, directly comparison will avoid hash collision.
FYI, at first the hash cache is not lazy evaluation, say the hash value is calculated at the constructor. There was many test timeout. After changing to the lazy evaluation, the timeout was gone.